src/metabase/driver/postgres.clj

```diff
@@ -298,6 +302,73 @@
   [driver [_ arg]]
   (sql.qp/->honeysql driver [:percentile arg 0.5]))
 
+(defn- datetime-diff-helper [x y unit]
+  (case unit
+    (:year :day)
+    (hx/cast
+     :integer
+     (hsql/call
+      :date_part
+      (hx/literal unit)
+      (hsql/call
+       (case unit :year :age :day :-)
+       (hsql/call :date_trunc (hsql/raw "'day'") y)
+       (hsql/call :date_trunc (hsql/raw "'day'") x))))
+
+    :week
+    (hx// (datetime-diff-helper x y :day) 7)
+
+    :month
+    (hx/cast
+     :integer
+     (hx/+
+      (hx/* 12 (datetime-diff-helper x y :year))
+      (hsql/call
+       :date_part
+       (hsql/raw "'month'")
+       (hsql/call
+        :age
+        (hsql/call :date_trunc (hsql/raw "'day'") y)
```

Comment:
Why don't you use the `date-trunc` function we already have in this namespace?

Reply:
`->timestamp` in this namespace uses `hx/cast-unless-type-in` and only adds a cast for things that aren't already a `timestamp`, `timestamptz`, or `date`.

```clj
(defn- ->timestamp [honeysql-form]
  (hx/cast-unless-type-in "timestamp" #{"timestamp" "timestamptz" "date"} honeysql-form))
```

So this shouldn't be adding any casts unless your `x` and `y` actually need it or you forgot to propagate that information somewhere.

I'm going to push back on your trying to avoid indirection here. It's better if we reuse code as much as possible for stuff like this. For example maybe we have some issue where we're doing unneeded casting in Postgres -- if that's handled by one function, then we just need to fix one function. Having a set of well-tested low-level stuff like `->timestamp` or `date-trunc` makes the higher-level stuff easier to write since you don't need to figure out the nuances and gotchas for stuff like that

Reply:
I see what you mean by the benefits of reuse, e.g. from [this change](https://github.com/metabase/metabase/commit/f31221d88831c2cd866479437cc5693d5ea67078) to date-trunc.

I've changed it in https://github.com/metabase/metabase/commit/71f2e30f0df16fe84a9eb932096ffd7a75a3af7c